File: heapsort.py
def heapsort(lst):
    makeListIntoHeap(lst)
    makeHeapIntoSortedList(lst)
 
def makeListIntoHeap(lst):
    #go from ]0, size] inserting into heap
    for i in range(1, len(lst)):
        goodPos = False
        index = i#upheap while it's not the root or at right place
        
        while (index != 0) and (not goodPos):#stop if (index == 0) or (goodPos)
            parent = int( (index - 1)/ 2)
            if lst[index] > lst[parent]:#if parent is smaller, swap
                swap(lst, index, parent)
                index = parent
            else:
                goodPos = True

def makeHeapIntoSortedList(lst):
    for size in range(len(lst)-1, 0, -1): #remove all elements and downheap
        swap(lst, 0, size)#swap first and last (removing biggest)
        goodPos = False
        inRange = True
        index = 0
        
        while not goodPos and inRange: #downheap
            leftChild = (2*index) + 1
            rightChild = (2*index) + 2
            
            if leftChild > (size-1):
                inRange = False
            elif rightChild > (size-1):
                biggestChild = leftChild
            else:
                biggestChild = leftChild if ( lst[leftChild] > lst[rightChild] ) else rightChild
            
            if inRange: 
                if lst[index] < lst[biggestChild]:
                    swap(lst, index, biggestChild)
                    index = biggestChild
                else:#index is bigger than its children
                    goodPos = True

def swap(lst, i, j):
    temp = lst[i]
    lst[i] = lst[j]
    lst[j] = temp

lst = [23,57,75,33,6,8,56]
heapsort(lst)
print "{0}".format(lst)

File: heapsort.java
import java.util.Arrays;

public class HeapSort {

	public static void heapSort(int[] lst){
		makeListIntoHeap(lst);
		makeHeapIntoSortedList(lst);
	}
	
	public static void makeListIntoHeap(int[] lst){
		//go from ]0,size] inserting into heap
		for(int i=1; i != lst.length; ++i){
			boolean goodPos = false;
			int index = i;
			//upheap while index is not root or in the right place
			
			//stop if (index == 0) or (goodPos)
			while( (index != 0) && !goodPos){
				int parent = (int)( (index - 1) / 2);
				if( lst[index] > lst[parent]){
					swap(lst, index, parent);
					index = parent;
				} else {
					goodPos = true;
				}
			}
		}
	}
	
	public static void makeHeapIntoSortedList(int[] lst){
		//remove all elements and downheap
		for(int i = lst.length-1; i != -1; --i){
			swap(lst, 0, i); //pop largest (put first element at the back)
			int index = 0;
			boolean goodPos = false;
			boolean inRange = true;
			
			while(!goodPos && inRange){
				int leftChild = (2*index) + 1;
				int rightChild = (2*index) + 2;
				int biggestChild = leftChild;
				
				if(leftChild > (i-1) ){			//index has no children
					inRange = false;
				} else if(rightChild > (i-1)){	//index has only leftChild
					biggestChild = leftChild;
				} else {						//index has both children
					biggestChild = (lst[leftChild] > lst[rightChild])? leftChild : rightChild;
				}
				
				if(inRange){
					if(lst[index] < lst[biggestChild]){
						swap(lst, index, biggestChild);
						index = biggestChild;
					} else {
						goodPos = true;
					}
				}
			}
		}
	}
	
	public static void swap(int[] lst, int i, int j){
		int temp = lst[i];
		lst[i] = lst[j];
		lst[j] = temp;
	}
	
	public static void main(String[] args) {
		int[] list = new int[]{23,57,75,33,6,8,56};
		heapSort(list);
		System.out.println( Arrays.toString(list) );
	}
}
